src/Microsoft.Management.Deployment/PackageCatalogReference.cpp

```diff
@@ -0,0 +1,82 @@
+        m_isComposite = true;
+    }
+    bool PackageCatalogReference::IsComposite()
+    {
+        return m_isComposite;
+    }
+    winrt::Microsoft::Management::Deployment::PackageCatalogInfo PackageCatalogReference::Info()
+    {
+        return m_info;
+    }
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Management::Deployment::ConnectResult> PackageCatalogReference::ConnectAsync()
+    {
+        co_return Connect();
+    }
+    winrt::Microsoft::Management::Deployment::ConnectResult PackageCatalogReference::Connect()
+    {
+        ::AppInstaller::ProgressCallback progress;
+        std::shared_ptr<::AppInstaller::Repository::ISource> source;
+        if (m_compositePackageCatalogOptions)
+        {
+            std::vector<std::shared_ptr<::AppInstaller::Repository::ISource>> remoteSources;
+
+            for (uint32_t i = 0; i < m_compositePackageCatalogOptions.Catalogs().Size(); ++i)
+            {
+                auto catalog = m_compositePackageCatalogOptions.Catalogs().GetAt(i);
+                winrt::Microsoft::Management::Deployment::implementation::PackageCatalogInfo* catalogInfoImpl = get_self<winrt::Microsoft::Management::Deployment::implementation::PackageCatalogInfo>(catalog.Info());
+                ::AppInstaller::Repository::SourceDetails sourceDetails = catalogInfoImpl->GetSourceDetails();
+                std::shared_ptr<::AppInstaller::Repository::ISource> remoteSource = ::AppInstaller::Repository::OpenSource(sourceDetails.Name, progress).Source;
+                remoteSources.emplace_back(std::move(remoteSource));
+            }
+            ::AppInstaller::Repository::CompositeSearchBehavior searchBehavior = GetRepositoryCompositeSearchBehavior(m_compositePackageCatalogOptions.CompositeSearchBehavior());
+           
+            std::shared_ptr<::AppInstaller::Repository::ISource> installedSource;
+            // Check if search behavior indicates that the caller does not want to do local correlation.
+            if (m_compositePackageCatalogOptions.CompositeSearchBehavior() != Microsoft::Management::Deployment::CompositeSearchBehavior::RemotePackagesFromRemoteCatalogs)
+            {
+                installedSource = ::AppInstaller::Repository::OpenPredefinedSource(::AppInstaller::Repository::PredefinedSource::Installed, progress);
+            }
+            
+            // Create the composite source.
+            source = ::AppInstaller::Repository::CreateCompositeSource(installedSource, remoteSources, searchBehavior);
+        }
+        else
+        {
+            winrt::Microsoft::Management::Deployment::implementation::PackageCatalogInfo* catalogInfoImpl = get_self<winrt::Microsoft::Management::Deployment::implementation::PackageCatalogInfo>(m_info);
+            ::AppInstaller::Repository::SourceDetails sourceDetails = catalogInfoImpl->GetSourceDetails();
+            source = ::AppInstaller::Repository::OpenSource(sourceDetails.Name, progress).Source;
```

Comment:
>source = ::AppInstaller::Repository::OpenSource(sourceDetails.Name, progress).Source; [](http://example.com/codeflow?start=12&length=85)

OpenSourceByDetails?

Reply:
There's a cross-package locking issue when using that that I need more time to debug.

Reply:
Should have been fixed by one of my changes.